CDLUC3/dmsp_aws_prototype
# Feature/api get dmps

**Changes**
- Uc3DmpDynamo gem
  - Updated to allow the initializer to accept an incoming table name to support calls against our Index Dynamo Table (temp replacement for OpenSearch)
- DmpIndexer Lambda function
  -  Make it a little more DRY
  - Added a few new attributes (was originally just the DMP PK):
    - `modified` date to support sorting
    - `title` to support sorting and search
    - `abstract` to support search
  -  Switch to use `PERSON_INDEX` and `AFFILIATION_INDEX` instead of `ORCID_INDEX` and `ROR_INDEX` to more accurately represent the fact that they can be based on emails and Crossref org/funder ids as well
-  Dynamo seed script
  - Removed old Augmenter records that are no longer used.
  - Update the `dynamo_seed` script to add a `"org_access_level": "all"` attribute to the provenance record. This allows the DMPTool to query for all Orgs. API integrations will use `"org_access_level": "restricted"` along with a list of approved ROR ids that will be stored in SSM
 
- Uc3DmpProvenance gem changes
  - Update to fetch the approved list of ROR ids from SSM for the system making the call and append it to the JSON fetched from the Dynamo table record
  - Will be adding the `org_access_level` flag and SSM for_list updates to the [Add Api Client PR](https://github.com/CDLUC3/dmsp_aws_prototype/tree/add-api-client) that sets up the pilot partner API access
- Uc3DmpId gem changes
  - Updated the `finder` class to support searching by affilaition, funder, and person. The results are combined so, for example, if the request asks for a specific org and funder the results set will only return results that include both the specified funder and specified org
- GetDmps API Lambda function changes
  - Updated to do a security check using the changes made to the Uc3DmpProvenance gem so that we allow a caller whose `org_access_level` is set to `all` can query any `org` and if not, they can only query RORs that are in their new SSM parameter `ror_list`. If they do not specify an org it will use all of the RORs that are in that SSM parameter
  - Updated to pass the `org`, `funder` and `owner` params through to the Uc3DmpId `finder`
  - Updated to support search by `title` and `abstract` 
  - Updated to sort by `title` or `modified` (default is by `modified` date `desc`)
  - Updated to allow `page` and `per_page` pagination (default is `page` 1 and `per_page` 25 with a max of 100)
  - Updated response to include `total_items`, `total_pages` along with `first`, `prev`, `next`, `last` (when appropriate) to help make it easier for callers to paginate through results

**Deploying**
- To deploy the DmpIndexer Lambda: `cd src/lambdas/indexers/dmp && ruby sam_build_deploy.rb [env] true true [log_level]`
- To deploy the gems and GetDmp Lambda changes: `cd src/sam && ruby sam_build_deploy.rb [env] true true [log_level]`
- Run the seed_dynamo script to update the DMPTool provenance record with the new `org_access_level ` attribute: `cd src/scripts && ./seed_dynamo [env] dmptool dmptool.org`
- Rebuild the Index table: `cd src/scripts && ruby recreate_dmp_index.rb [env] [index table name]`
- Don't forget to run Sceptre updates for each of the API Clients to add their ROR lists to SSM. These can be found in the [Add Api Client PR](https://github.com/CDLUC3/dmsp_aws_prototype/tree/add-api-client)

**Testing**
To test the changes to the DmpIndexer Lambda
- Log into the AWS console
- Open the DynamoDB Index table
- Click 'Explore table items' and select one of the records whose PK is a DMP ID (e.g. `DMP#doi.org/10.12345/A1b348ygh354`)
- Make note of the `affiliation_ids`, `funder_ids`, `people_ids` and emails in `people` in the above record.
- For each of the `affiliation_ids`, query the same table for `"PK": "AFFILIATION_INDEX" and "SK": [the id]` and then make sure the `dmps` list includes the DMP ID in the original record. It should also have the `modified`, `title` and `abstract` values.
- For each of the `funder_ids`, query the same table for `"PK": "FUNDER_INDEX" and "SK": [the id]` and then make sure the `dmps` list includes the DMP ID in the original record. It should also have the `modified`, `title` and `abstract` values.
- For each of the `people_ids` AND `emails`, query the same table for `"PK": "PERSON_INDEX" and "SK": [the id]` and then make sure the `dmps` list includes the DMP ID in the original record. It should also have the `modified`, `title` and `abstract` values. 

To test the changes to the API
- Authenticate as the DMPTool to get a token. For example: `curl -v -H 'Authorization: Basic [base64 encoded crews]' -H 'Content-Type: application/x-www-form-urlencoded' -d 'grant_type=client_credentials&scope=https://auth.dmphub.uc3[env].cdlib.net/[env].read' "https://auth.dmphub.uc3[env].cdlib.net/oauth2/token"`
- Make note of the token returned by the above and use it in all subsequent requests (e.g. 
- Search without filters for ALL dmps: `curl -v -H 'Accept: application/json' -H 'Authorization: Bearer [token]' "https://api.dmphub.uc3[env].cdlib.net/dmps"
- Search for dmps for a specific org (e.g. UC Riverside): `curl -v -H 'Accept: application/json' -H 'Authorization: Bearer [token]' "https://api.dmphub.uc3[env].cdlib.net/dmps?org=03nawhv43"
- Search for dmps for a specific funder (e.g. NIH): `curl -v -H 'Accept: application/json' -H 'Authorization: Bearer [token]' "https://api.dmphub.uc3[env].cdlib.net/dmps?funder=10.13039/100000002"
- Search for dmps for your own email address: `curl -v -H 'Accept: application/json' -H 'Authorization: Bearer [token]' "https://api.dmphub.uc3[env].cdlib.net/dmps?owner=[email with '%40' instead of the '@']"
- Search for dmps for your own ORCID: `curl -v -H 'Accept: application/json' -H 'Authorization: Bearer [token]' "https://api.dmphub.uc3[env].cdlib.net/dmps?owner=[orcid without the domain like 0000-0000-0000-0000]" 
- Search for combos of the above to make sure that they filter properly. For example if you specify UC riverside as the org and NIH as the funder make sure it doesn't include DMPs funded by NSF.
- Include a `search=[term]` to make sure it further filters the results by your search term
- Test the sort by `sort=title` and `sort=modified` and `sort_dir=[asc | desc]`
- Test the pagination `page=[n]` and `per_page=[n]`
